### PR TITLE
Full dynamic tracing using capstone disasm

### PR DESCRIPTION
**Pull request by ParkHanbum**
You can check what purpose of this PR by follow below steps.
1. modify misc/debug-mcount.cmd
```
  file uftrace
  set breakpoint pending on

  #b command_record
  catch exec
  commands
  #  set follow-fork-mode child
   b *main+26
    continue
  end

 r record -L. -vv -d xxx --keep-pid -P. tests/t-abc
```

2. build target
```
$ make test TESTARG='212'
```

3. run gdb with debug.sh
```
$ sh misc/debug.sh
```
4. now you reach here
```
(gdb) disas main
Dump of assembler code for function main:
(gdb) disas
Dump of assembler code for function main:
   0x0000000000400420 <+0>:     callq  0x400ff0
   0x0000000000400425 <+5>:     nop
   0x0000000000400426 <+6>:     push   %rbx
   0x0000000000400427 <+7>:     push   %rcx
   0x0000000000400428 <+8>:     jle    0x400438 <main+24>
   0x000000000040042a <+10>:    mov    0x8(%rsi),%rdi
   0x000000000040042e <+14>:    callq  *0x200bc4(%rip)        # 0x600ff8
   0x0000000000400434 <+20>:    mov    %eax,%ebx
   0x0000000000400436 <+22>:    jmp    0x40043a <main+26>
   0x0000000000400438 <+24>:    xor    %ebx,%ebx
=> 0x000000000040043a <+26>:    callq  0x40055b <a>
```
5. step into function a to check what different. 
```
(gdb) disas
Dump of assembler code for function a:
=> 0x000000000040055b <+0>:     callq  0x400ff0
   0x0000000000400560 <+5>:     nop
   0x0000000000400561 <+6>:     nop

(gdb) si
0x0000000000400ff0 in ?? ()
(gdb) x/i $rip
=> 0x400ff0:    jmpq   *0x2(%rip)        # 0x400ff8
(gdb) si
__dentry__ () at /home/m/git/uftrace/arch/x86_64/dynamic_hook.S:46
46              sub $48, %rsp
```
6. call dynamic_entry in __dentry__
first, just call function mcount_entry in dynamic_entry.
```
80              call dynamic_entry
(gdb) si
dynamic_entry (parent_loc=0x7fffffffe2d8, child=4195680, child_loc=0x7fffffffe2d0, regs=0x7fffffffe2a0) at /home/m/git/uftrace/libmcount/mcount.c:986
986             result = mcount_entry(parent_loc, child, regs);
```
after that, find location that stored origin code.
and change return address of current function to make run origin code. 
```
uintptr_t origin_code_addr = mcount_find_origin_code_addr(child);
*child_loc = (unsigned long)origin_code_addr;

(gdb) p/x origin_code_addr
$3 = 0x947360
(gdb) x/10i 0x947360
   0x947360:    push   %rbp
   0x947361:    mov    %rsp,%rbp
   0x947364:    nop
   0x947365:    jmpq   *0x0(%rip)        # 0x94736b
   0x94736b:    (bad)
   0x94736c:    add    $0x40,%eax
   0x947371:    add    %al,(%rax)
   0x947373:    add    %al,(%rax)
   0x947375:    add    %al,(%rax)
   0x947377:    add    %dh,(%rcx)
(gdb) x/gx 0x94736b
0x94736b:       0x0000000000400560
(gdb) x/i 0x400560
   0x400560 <a+5>:      nop
```

continue step-in until reach here. you can see.
```
__dentry__ () at /home/m/git/uftrace/arch/x86_64/dynamic_hook.S:98
98              retq
(gdb) x/gx $rsp
0x7fffffffe2d0: 0x0000000000947360
(gdb) si
0x0000000000947360 in ?? ()
(gdb) x/10i $rip
=> 0x947360:    push   %rbp
   0x947361:    mov    %rsp,%rbp
   0x947364:    nop
   0x947365:    jmpq   *0x0(%rip)        # 0x94736b
   0x94736b:    (bad)
   0x94736c:    add    $0x40,%eax
(gdb) x/gx 0x94736b
0x94736b:       0x0000000000400560
(gdb) si
0x0000000000947361 in ?? ()
(gdb)
0x0000000000947364 in ?? ()
(gdb)
0x0000000000947365 in ?? ()
(gdb)
0x0000000000400560 in a ()
(gdb) x/i $rip
=> 0x400560 <a+5>:      nop
(gdb) x/5i 0x40055b
   0x40055b <a>:        callq  0x400ff0
=> 0x400560 <a+5>:      nop
```
as you can see, it return to location of origin code. 

OK testcases increased by adding testcases for 'full-dynamic'

```
211 replay_filter_C     : OK OK OK OK OK OK OK OK OK OK
212 full_dyanmic        : SG SG SG SG SG SG SG SG SG SG

runtime test stats
====================
total  2120  Tests executed (success: 81.75%)
  OK:  1670  Test succeeded
  OK:    63  Test succeeded (with some fixup)
```

```
212 full_dyanmic        : OK OK OK OK OK OK OK OK OK OK
runtime test stats
====================
total  2120  Tests executed (success: 82.22%)
  OK:  1680  Test succeeded
  OK:    63  Test succeeded (with some fixup)
```

### CODE REVIEW

**Comment by namhyung** [comment]
Thanks for doing this.  I'll take a look at it later..

**Comment by ParkHanbum** [comment]
@namhyung this PR has been updated.

**Comment by honggyukim** [comment]
Hi @ParkHanbum, it looks a great work. But I feel lost what to do with this PR. Could you please write how I can reproduce and test your work?

File: configure [hunk 1]
```diff
@@ -170,6 +171,7 @@ print_feature()
 printf "...%15s: %s\n" "prefix" "${prefix}"
 print_feature "libelf" "have_libelf" "more flexible ELF data handling"
 print_feature "libdw" "have_libdw" "DWARF debug info support"
+print_feature "libcapstone" "have_libcapstone" "Disassembler to full-dynamic"
```
Comment (namhyung) [inline configure]: Please put this in the last line with message as "full dynamic tracing support".

File: arch/x86_64/mcount-regs.h [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+	/* floating-point registers */
+	X86_REG_XMM0,
+	X86_REG_XMM1,
+	X86_REG_XMM2,
+	X86_REG_XMM3,
+	X86_REG_XMM4,
+	X86_REG_XMM5,
+	X86_REG_XMM6,
+	X86_REG_XMM7,
+};
```
Comment (namhyung) [inline arch/x86_64/mcount-regs.h]: As I said earlier, I'd like to rename it to reduce possible confusion.

File: arch/x86_64/mcount-dynamic.c [hunk 3]
```diff
@@ -295,11 +310,482 @@ static int update_xray_func(struct mcount_dynamic_info *mdi, struct sym *sym)
...
+
+	for(i = 0;i < count;i++) {
+		int res = check_instrumentable(csh_handle, CS_MODE_64, &insn[i]);
+		if (res & CODE_PATCH_NO) {
+			pr_dbg("\tThe instruction not supported : %s\t %s\n",
+				insn[i].mnemonic, insn[i].op_str);
+			return INSTRUMENT_SKIPED;
+		}
+		code_size += insn[i].size;
+		if (code_size > g_patch_code_size) {
```
Comment (namhyung) [inline arch/x86_64/mcount-dynamic.c]: I think we can stop if `code_size == g_patch_code_size` too.

File: libmcount/mcount.c [hunk 4]
```diff
@@ -967,6 +967,39 @@ int mcount_entry(unsigned long *parent_loc, unsigned long child,
...
+{
+	int result;
+	result = mcount_entry(parent_loc, child, regs);
+	if (!result) {
+		/*
+		 * dynamic_entry returns the address
+		 * holding the patched original code.
+		 */
+		uintptr_t origin_code_addr = mcount_find_origin_code_addr(child);
+		return origin_code_addr;
```
Comment (namhyung) [inline libmcount/mcount.c]: You can update it here like:

```c
*parent_loc = origin_code_addr;
````

File: arch/x86_64/mcount-dynamic.c [hunk 5]
```diff
@@ -295,11 +310,482 @@ static int update_xray_func(struct mcount_dynamic_info *mdi, struct sym *sym)
...
+				break;
+		}
+	}
+	return status;
+}
+
+
+#define INSTRUMENT_SKIPED			-2
+#define INSTRUMENT_FAILED			-1
+#define INSTRUMENT_SUCCESS			 0
```
Comment (namhyung) [inline arch/x86_64/mcount-dynamic.c]: Move this somewhere the generic dynamic layer (libmcount/dynamic.c) can share.

**Comment by ParkHanbum** [comment]
@honggyukim I have updated comments. would review about it, please?

**Comment by honggyukim** [comment]
@ParkHanbum Okay. I will test it later. Thanks!

File: tests/t212_full_dynamic.py [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+
+    def pre(self):
+        if platform.machine().startswith('arm'):
+            return TestBase.TEST_SKIP
+        return TestBase.TEST_SUCCESS
+
+    def build(self, name, cflags='', ldflags=''):
+        cflags = cflags.replace('-pg', '')
+        cflags = cflags.replace('-finstrument-functions', '')
+        cflags += ' -fno-pie -fno-plt'  # workaround of build failure
```
Comment (honggyukim) [inline tests/t212_full_dynamic.py]: Hmm.. why do we need these flags?

**Comment by honggyukim** [comment]
@ParkHanbum I just summarized the execution trace of this patch so make it simpler. I will see the details later on again. It'd be good to have more explanation on each function. Thanks for your work!

**1.  startup time setup**
```diff
 mcount_startup() {
   mcount_dynamic_update() {
+    disassembler_init();
     do_dynamic_update() {
       mcount_patch_func() {
         patch_fentry_func();
+        instrument() {
+          check_instrument_size() {
+            cs_disasm();
+            check_instrumentable();
+          }
+          do_instrument(){
+            get_dentry_rel_trampoline_addr() {
+              get_trampoline_addr();
+            }
+            patch_code();
+          } /* do_instrument */
+        } /* instrument */
       } /* mcount_patch_func */
     } /*do_dynamic_update */
+    disassembler_fini();
   } /* mcount_dynamic_update() */
 } /* mcount_startup */
```

**2. execution on function entry**
```diff
+GLOBAL(__dentry__) {
+  dynamic_entry() {
     mcount_entry();
+    mcount_find_origin_code_addr();
+  } /* dynamic_entry */
 } /* __dentry__ */
```

**Comment by ParkHanbum** [comment]
@honggyukim thank you for your review!!! gracious!
different result is normal when testing s-signal.c.
If you run it with the -v option, you'll know why.

and now, there is some problem while load the libmcount.so. I am determining the cause of the problem.

**Comment by honggyukim** [comment]
@ParkHanbum Right. I found that the hidden functions are skipped because their function size is too small to patch a function call to `__dentry__`.

I have one more thing to ask. I think we'd better have one more option that sets the minimum function size that can be traced instead of just using `g_patch_code_size`. We can focus only on large functions ignoring small functions with it. The minimum size has to be greater than `g_patch_code_size`, but it can be a lot larger than that if the user wants to trace only big functions.

**Comment by ParkHanbum** [comment]
@honggyukim sure, it seems not so difficult . I will update immediately after this PR is merged. 
thank you for suggestion

**Comment by honggyukim** [comment]
I have one more idea. This feature doesn't work when the target binary is stripped because it doesn't know the addresses of each function as follows:
```
$ gcc s-abc.c

$ uftrace -P. a.out
some functions cannot be patched dynamically
# DURATION     TID     FUNCTION
            [ 66776] | main() {
            [ 66776] |   a() {
            [ 66776] |     b() {
            [ 66776] |       c() {
   1.537 us [ 66776] |         getpid();
   6.677 us [ 66776] |       } /* c */
   7.440 us [ 66776] |     } /* b */
   8.193 us [ 66776] |   } /* a */
   9.717 us [ 66776] | } /* main */

# do strip out the symbol info
$ strip a.out

# try full dynamic tracing again
$ uftrace -P. a.out
. cannot be patched dynamically
# DURATION     TID     FUNCTION
   1.607 us [ 66813] | getpid();
```

But I would like to have a way to pass the symbol information separately. For example, once we have symbol information, we extract (maybe using `nm`) symbol info and save it as a file, let's say `nm.sym`. Then it can be stripped.

The stripped binary can be dynamically traced based on the `nm.sym`.

So I would like to have below execution sequence.
```
$ gcc s-abc.c

# extract symbol info and save it
$ nm a.out > nm.sym

# do strip out the symbol info
$ strip a.out

# try full dynamic tracing again with an external symbol info from nm.sym
$ uftrace -P. --patch-symbol=nm.sym a.out
some functions cannot be patched dynamically
# DURATION     TID     FUNCTION
            [ 66776] | main() {
            [ 66776] |   a() {
            [ 66776] |     b() {
            [ 66776] |       c() {
   1.537 us [ 66776] |         getpid();
   6.677 us [ 66776] |       } /* c */
   7.440 us [ 66776] |     } /* b */
   8.193 us [ 66776] |   } /* a */
   9.717 us [ 66776] | } /* main */
```

This can greatly increase the usability quite a lot.

**Comment by honggyukim** [comment]
The below `T` and `t` symbols have to be parsed before tracing begins.
```
$ gcc s-abc.c
$ nm a.out | grep " T \| t "
0000000000400566 t a
0000000000400577 t b
0000000000400588 t c
00000000004004a0 t deregister_tm_clones
0000000000400520 t __do_global_dtors_aux
0000000000600e18 t __do_global_dtors_aux_fini_array_entry
0000000000400674 T _fini
0000000000400540 t frame_dummy
0000000000600e10 t __frame_dummy_init_array_entry
0000000000400400 T _init
0000000000600e18 t __init_array_end
0000000000600e10 t __init_array_start
0000000000400670 T __libc_csu_fini
0000000000400600 T __libc_csu_init
00000000004005b7 T main
00000000004004e0 t register_tm_clones
0000000000400470 T _start
```

**Comment by honggyukim** [comment]
I tried it for original `clang` and it seems the record is kind of normally done, but shows some error messages.
```
$ uftrace record -P. -t 10ms ./clang hello.c
some functions cannot be patched dynamically
some functions cannot be patched dynamically
clang-6.0: error: unable to execute command: Segmentation fault (core dumped)
clang-6.0: error: linker command failed due to signal (use -v to see invocation)
```

**Comment by ParkHanbum** [comment]
this PR useless until fix the issue #619. so, closed this PR.
for other developer who want to develop this feature.

**Pull request closed without merging**